### PR TITLE
DB-based light client backend

### PR DESCRIPTION
**Pull request by svyatonik**
This PR is about switch from using in-memory to database backend on light nodes. Additionally, `client/src/light.rs` has been split into separate files in `client/src/light/` subfolder, because it grew too big.

Databases of full && light nodes are made incompatible (i.e. you can't open full db by light client and light db by full client). Initially tried to use the same db for both nodes, but found this a bad decision, because it:
1) would require upgrading full db when light db structure changes and vice versa;
2) lead to errors on full clients when we're trying to use light node' database (missing body, etc).

P.S.: syncing with `--light` is broken on current master, I'm currently working on it, but still wanted to start reviews of my previous work.

### CODE REVIEW

**Comment by svyatonik** [comment]
deadlock is fixed in #260

File: substrate/state-machine/src/lib.rs [hunk 1]
```diff
@@ -227,7 +227,7 @@ pub fn prove<B: TryIntoTrieBackend, Exec: CodeExecutor>(
 }
 
 /// Check execution proof, generated by `prove` call.
-pub fn proof_check<Exec: CodeExecutor>(
+pub fn execution_proof_check<Exec: CodeExecutor>(
```
Comment (pepyakin) [inline substrate/state-machine/src/lib.rs]: in comment: generated by `prove_execution` call

File: substrate/client/src/in_mem.rs [hunk 2]
```diff
@@ -108,16 +103,17 @@ impl<Block: BlockT> Blockchain<Block> {
...
 		Blockchain {
-			storage: RwLock::new(
-				BlockchainStorage {
-					blocks: HashMap::new(),
-					hashes: HashMap::new(),
-					best_hash: Default::default(),
-					best_number: Zero::zero(),
-					genesis_hash: Default::default(),
-				})
+			storage: storage.clone(),
```
Comment (pepyakin) [inline substrate/client/src/in_mem.rs]: Is this clone really needed?

File: substrate/client/src/light/backend.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+use error::{Error as ClientError, ErrorKind as ClientErrorKind, Result as ClientResult};
+use light::blockchain::{Blockchain, Storage as BlockchainStorage};
+use light::fetcher::Fetcher;
+
+/// Light client backend.
+pub struct Backend<S, F> {
+	blockchain: Arc<Blockchain<S, F>>,
+}
+
+/// Ligh block (header and justification) import operation.
```
Comment (pepyakin) [inline substrate/client/src/light/backend.rs]: s/Ligh/Light

File: substrate/client/src/in_mem.rs (hunk 2 above)
Comment (svyatonik) [reply substrate/client/src/in_mem.rs]: Not in this PR :) It is actually from the next PR (https://github.com/paritytech/polkadot/pull/251/files#diff-8f118ee188780f9183b6af320625cfb9R106)

**Pull request merged**